Add grid force

ccccclw opened the pull request:
The grid force is added on reference and CUDA platform and associated Python module is modified accordingly. There is a test system under `test_density`, while forces seem to be calculated correctly, the fitting process is much slower than what is implemented in the older version of MELD (blue vs orange in the plot below). Toy system force tests are also added and passed.  As I continue to find out where goes wrong, could you check if the python part is written correctly? 

![image](https://user-images.githubusercontent.com/49604002/144141094-00d59e52-b811-46d7-ad61-e0141b30bba9.png)

jlmaccal commented on meld/runner/transform/restraints/meld/meld.py:
```diff
@@ -25,6 +25,7 @@
 from simtk.openmm import app  # type: ignore
 
 import numpy as np  # type: ignore
+import scipy.ndimage
```
```suggestion
import scipy.ndimage  # type: ignore
```

jlmaccal commented on meld/runner/transform/restraints/meld/meld.py:
```diff
@@ -76,14 +77,14 @@ def add_interactions(
-            for density in self.density_manager.densities:
-                self.tracker.add_density(density)
-                blurred = _compute_density_potential(density)
+            for index,density in enumerate(self.density_manager.densities):
```
```suggestion
            for index, density in enumerate(self.density_manager.densities):
```

jlmaccal commented on meld/runner/transform/restraints/meld/meld.py:
```diff
@@ -76,14 +77,14 @@ def add_interactions(
-                blurred = _compute_density_potential(density)
+            for index,density in enumerate(self.density_manager.densities):
+                self.tracker.add_density(index, density,0)
+                blurred = _compute_density_potential(density,0,origin=True)
```
```suggestion
                blurred = _compute_density_potential(density, 0, origin=True)
```

jlmaccal commented on plugin/platforms/cuda/tests/TestCudaMeldForce.cpp:
```diff
@@ -1131,23 +1197,24 @@ int main(int argc, char* argv[]) {
-        testMultipleGroups();
-        testBigSystem();
-        testPBC();
+        // testDistRest();
+        // testTorsRest();
+        // testDistProfileRest();
+        // testTorsProfileRest();
+        // testDistRestChangingParameters();
+        // testHyperbolicDistRest();
+        // testGMMRest1Pair1Component();
+        // testGMMRest1Pair1Component0Scale();
+        // testGMMRest1Pair2Component();
+        // testGMMRest2Pair2Component();
+        // testGMMRest3Pair2Component();
+        testGridPotentialRest();
+        // testGroupSelectsCorrectly();
+        // testCollectionSelectsCorrectly();
+        // testSingleGroup();
+        // testMultipleGroups();
+        // testBigSystem();
+        // testPBC();
```
Please uncomment the other tests.

jlmaccal commented on plugin/platforms/reference/tests/TestReferenceMeldForce.cpp:
```diff
@@ -778,6 +778,72 @@ void testGMMRest3Pair2Component() {
+    context.setPositions(positions);
+    State stateI = context.getState(State::Energy | State::Forces);
+    
+    //  V_{xyz} = atom_mass * (1/L^3)(V_{000}(x_max-x)(y_max-y)(z_max-z) + 
+    //            V_{100}x(y_max-y)(z_max-z) + 
+    //            V_{010}(x_max-x)y(z_max-z) + 
+    //            V_{001}(x_max-x)(y_max-y)z + 
+    //            V_{101}x(y_max-y)z + 
+    //            V_{011}(x_max-x)yz + 
+    //            V_{110}xy(z_max-z) + 
+    //            V_{111}xyz)  (L is the grid length)
+    //  V_total = V_0 + V_1 = 0.188 + 0.352 = 0.54
+    //  F_0 = -(dV/dx,dV/dy,dV/dz) = (V_{011}yz/L, V_{101}xz/L, V_{110}xy/L) = (0.4,0.3,0.24)
+    // cout << stateI.getPotentialEnergy() << endl;
+    // cout << stateI.getForces()[0] << endl;
+    // cout << stateI.getForces()[1] << endl;
```
Can you remove this if it is no longer used?

jlmaccal replied:
Please finish implementing this test. It should either pass or fail. It's usually a good idea to explain what you expect the value to be and why.

jlmaccal commented on plugin/platforms/reference/tests/TestReferenceMeldForce.cpp:
```diff
@@ -1129,9 +1195,9 @@ void testPBC() {
-        testDistRest();
-        testTorsRest();
-        testDistRestChangingParameters();
+        // testDistRest();
+        // testTorsRest();
+        // testDistRestChangingParameters();
```
Please uncomment the other tests.

jlmaccal commented on plugin/platforms/cuda/tests/TestCudaMeldForce.cpp:
```diff
@@ -778,6 +778,72 @@ void testGMMRest3Pair2Component() {
+    // cout << stateI.getPotentialEnergy() << endl;
+    // cout << stateI.getForces()[0] << endl;
+    // cout << stateI.getForces()[1] << endl;
+    float expectedEnergy = 0.54;
+    Vec3 expectedForce1 = Vec3(0.4,0.3,0.24);
+    Vec3 expectedForce2 = Vec3(-0.8,-0.96,-1.2);
```
Could you add a comment explaining where these values come from and how you know they are correct?

ccccclw replied:
It comes from the calculation shown here.
https://github.com/maccallumlab/meld/blob/b39fe6ebf9701ef128e6741e3e1c165c347e4739/plugin/platforms/reference/tests/TestReferenceMeldForce.cpp#L827

jlmaccal replied:
OK, could you move that comment over to this file as well? It will just help in the future if each test is self documenting.

jlmaccal reviewed (commented):
A couple of other points to address:

- Please move the `test_density` files into a separate repository. It's great to have those for reference, but they add a few MB to the source tree.
- Please run the `black` code formatter on all of the python files that changed to make the formatting consistient with the existing MELD tree. You can easily run this in vscode, or from the command line.

jlmaccal commented:
I can't see what the x-axis is in the graph to see what sort of timescale we are looking at. It's also not clear to me exactly what we changed between the old and new version.

One thing that would be useful is some of the diagnostics from `analyze_remd`, specifically the `visualize_trace`, `visualize_alpha`, and `visualize_accept` sub-commands. Those will give use diagnostics about how well replica exchange i adapting / mixing.

jlmaccal commented:
Also, you can see below that the tests are not passing. I think this is just the missing `type: ignore` comment that I suggested in the code review.

jlmaccal commented on test_density/setup_emap.py:
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
+# encoding: utf-8
+import numpy as np
+import meld
+from meld.comm import MPICommunicator
+from meld import comm, vault
+import mdtraj as md
+from meld import system
+from meld.remd import ladder, adaptor, leader
+from meld.system.scalers import LinearRamp,ConstantRamp
```
Almost all of these should now be improted into the base `meld` module. So you should only need `import meld` and them `meld.LinaerRamp`, etc. It's not a big deal, but it cleans up the imports a lot.

jlmaccal commented on test_density/setup_emap.py:
```diff
@@ -0,0 +1,186 @@
+    s = b.build_system([p])
+    s.temperature_scaler = meld.ConstantTemperatureScaler(300.0 * u.kelvin)  #
+    n_res = s.residue_numbers[-1]
+    blur_scaler = s.restraints.create_scaler(
+        "linear_blur", alpha_min=0, alpha_max=1, min_blur=0.0, max_blur=0.0
+    )
```
This doesn't seem to be doing any blurring?

ccccclw replied:
Yes, this is a test system that doesn't require blur to fit, I was trying to compare the result with previous code under the same set up.

jlmaccal replied:
I see.

jlmaccal commented:
> I can't see what the x-axis is in the graph to see what sort of timescale we are looking at. It's also not clear to me exactly what we changed between the old and new version.

So, dark mode was making it so I couldn't see the x-axis. But what units is it in? Could it be that the force is just weaker that it was previously, or do you think something else is wrong?

ccccclw commented:
> > I can't see what the x-axis is in the graph to see what sort of timescale we are looking at. It's also not clear to me exactly what we changed between the old and new version.
> 
> So, dark mode was making it so I couldn't see the x-axis. But what units is it in? Could it be that the force is just weaker that it was previously, or do you think something else is wrong?

The x-axis unit is ns. I was expecting the old and new version behave similarly, but the old one fit finished within ~50ns (similar to the test on the same system shown by NAMD) but the newer one took much longer. I checked the parameters but haven't found what caused this, will need to investigate a bit more. One thing I am not sure is that if the code I wrote correctly for density restraints here, e.g. are the restraints added and updated correctly? https://github.com/maccallumlab/meld/blob/b39fe6ebf9701ef128e6741e3e1c165c347e4739/meld/runner/transform/restraints/meld/meld.py

jlmaccal commented:
e similarly, but the old one fit finished within ~50ns (similar to the test on the same system shown by NAMD) but the newer one took much longer. I checked the parameters but haven't found what caused this, will need to investigate a bit more. One thing I am not sure is that if the code I wrote correctly for density restraints here, e.g. are the restraints added and updated correctly? https://github.com/maccallumlab/meld/blob/b39fe6ebf9701ef128e6741e3e1c165c347e4739/meld/runner/transform/restraints/meld/meld.py

Nothing jumped out at me as being incorrect when I looked at that part. It does seem to be working, but perhaps the scaling of the energy is different now? I also found the linear progression of the blue curve to be a bit strange. One thing you could do would be to write out the parameters and values of the grid for both the new and old versions.

jlmaccal commented:
Or alternatively, create an extremely simple test system with a single atom and a Guassian density or something like that. With such a complex system, it's often difficult to get a handle on what is going wrong. Simpler is better.
